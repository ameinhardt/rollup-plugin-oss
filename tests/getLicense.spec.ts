import { getLicenses } from '../index.js';

const possibleValues = [
  {
    type: 'ISC',
    url: 'https://opensource.org/licenses/ISC'
  },
  42,
  'MIT OR GPL-2.0 AND ISC',
  'BSD-3-Clause',
  '(MIT AND (LGPL-2.1+ AND BSD-3-Clause WITH GPL-3.0-linking-exception))',
  '(ISC OR (GPL-2.0+))',
  'See LICENSE.md'
];

describe('gets licenses from package attribute', function() {
  for (const [index, value] of possibleValues.entries()) {
    it(`#${index + 1} ${typeof value === 'string' ? value : 'object' }`, function() {
      try {
        const licenses = getLicenses(value as string | Record<string, string>);
        expect(licenses).toMatchSnapshot();
      } catch (error) {
        expect(error).toMatchSnapshot();
      }
    });
  }
});
