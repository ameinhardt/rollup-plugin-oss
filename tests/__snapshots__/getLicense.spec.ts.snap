// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gets licenses from package attribute #1 object 1`] = `
{
  "infos": [
    {
      "license": "ISC",
    },
  ],
  "license": "ISC",
}
`;

exports[`gets licenses from package attribute #2 object 1`] = `[TypeError: can't parse license]`;

exports[`gets licenses from package attribute #3 MIT OR GPL-2.0 AND ISC 1`] = `[Error: can't unambiguously parse license definition]`;

exports[`gets licenses from package attribute #4 BSD-3-Clause 1`] = `
{
  "infos": [
    {
      "license": "BSD-3-Clause",
    },
  ],
  "license": "BSD-3-Clause",
}
`;

exports[`gets licenses from package attribute #5 (MIT AND (LGPL-2.1+ AND BSD-3-Clause WITH GPL-3.0-linking-exception)) 1`] = `
{
  "infos": [
    {
      "exception": "GPL-3.0-linking-exception",
      "license": "BSD-3-Clause",
    },
    {
      "license": "LGPL-2.1",
      "plus": true,
    },
    {
      "license": "MIT",
    },
  ],
  "license": "(MIT AND (LGPL-2.1+ AND BSD-3-Clause WITH GPL-3.0-linking-exception))",
}
`;

exports[`gets licenses from package attribute #6 (ISC OR (GPL-2.0+)) 1`] = `[Error: found multiple license options]`;

exports[`gets licenses from package attribute #7 See LICENSE.md 1`] = `[Error: can't parse license definition]`;
